File: sty_algorithm/src/main/java/com/sty/algorithm/leetcode/editor/cn/[242]有效的字母异位词.java
package com.sty.algorithm.leetcode.editor.cn;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicBoolean;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {

    public boolean isAnagram(String s, String t) {
        AtomicBoolean res = new AtomicBoolean(true);

        if (s.length() != t.length()) {
            return false;
        }

        Map<Character, Integer> sMap = f(s);
        Map<Character, Integer> tMap = f(t);

        sMap.forEach((k, v) -> {
            if (!tMap.containsKey(k)) {
                res.set(false);
                return;
            }
            if (!Objects.equals(tMap.get(k), v)) {
                res.set(false);
            }
        });

        return res.get();
    }

    private Map<Character, Integer> f(String s) {
        char[] chars = s.toCharArray();
        Map<Character, Integer> map = new HashMap<>();

        for (char c : chars) {
            if (map.containsKey(c)) {
                Integer v = map.get(c);
                map.put(c, v + 1);
            } else {
                map.put(c, 0);
            }
        }

        return map;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: sty_algorithm/src/main/scala/com/sty/algorithm/leetcode/editor/cn/[242]有效的字母异位词.scala
import scala.collection.mutable

//leetcode submit region begin(Prohibit modification and deletion)
object Solution {

    def isAnagram(s: String, t: String): Boolean = {
        if (s.length != t.length) {
            return false
        }

        val sMap = fun(s)
        val tMap = fun(t)

        var res = true

        for (elem <- sMap) {
            if (!tMap.contains(elem._1)) {
                res = false
                return res
            }
            if (tMap(elem._1) != elem._2) {
                res = false
                return res
            }
        }

        res

    }

    def fun(s: String) = {
        val array = s.toCharArray
        val map: mutable.HashMap[Char, Int] =  mutable.HashMap()
        array.foreach(c => {
            if (map.contains(c)) {
                val v = map.get(c)
                map.put(c, v.get + 1)
            } else {
                map.put(c, 0)
            }
        })
        map
    }
}
//leetcode submit region end(Prohibit modification and deletion)
